internal/sbi/httpcallback/api_handle_dereg_notification.go

```diff
@@ -0,0 +1,148 @@
+package httpcallback
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+
+	amf_context "github.com/free5gc/amf/internal/context"
+	"github.com/free5gc/amf/internal/logger"
+	"github.com/free5gc/amf/internal/sbi/consumer"
+	"github.com/free5gc/openapi"
+	"github.com/free5gc/openapi/Nudm_UEContextManagement"
+	"github.com/free5gc/openapi/models"
+)
+
+func HTTPAmfHandleDeregistrationNotification(c *gin.Context) {
+	// TS 23.502 - 4.2.2.2.2 - step 14d
+	logger.CallbackLog.Infoln("Handle Deregistration Notification")
+
+	var deregData models.DeregistrationData
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		logger.CallbackLog.Errorf("Get Request Body error: %+v", err)
+		problemDetails := models.ProblemDetails{
+			Title:  "System failure",
+			Status: http.StatusInternalServerError,
+			Detail: err.Error(),
+			Cause:  "SYSTEM_FAILURE",
+		}
+		c.JSON(http.StatusInternalServerError, problemDetails)
+		return
+	}
+
+	err = openapi.Deserialize(&deregData, requestBody, "application/json")
+	if err != nil {
+		problemDetails := "[Request Body] " + err.Error()
+		rsp := models.ProblemDetails{
+			Title:  "Malformed request syntax",
+			Status: http.StatusBadRequest,
+			Detail: problemDetails,
+		}
+		logger.CallbackLog.Errorln(problemDetails)
+		c.JSON(http.StatusBadRequest, rsp)
+		return
+	}
+
+	ueid := c.Param("ueid")
+	amfSelf := amf_context.GetSelf()
+	ue, ok := amfSelf.AmfUeFindByUeContextID(ueid)
+	if !ok {
+		logger.CallbackLog.Errorf("AmfUe Context[%s] not found", ueid)
+		problemDetails := &models.ProblemDetails{
+			Status: http.StatusNotFound,
+			Cause:  "CONTEXT_NOT_FOUND",
+		}
+		c.JSON(http.StatusNotFound, problemDetails)
+		return
+	}
+
+	switch deregData.DeregReason {
+	case models.DeregistrationReason_UE_INITIAL_REGISTRATION:
+		// TS 23.502 - 4.2.2.2.2 General Registration
+		// Invokes the Nsmf_PDUSession_ReleaseSMContext for the corresponding access type
+		ue.SmContextList.Range(func(key, value interface{}) bool {
+			smContext := value.(*amf_context.SmContext)
+
+			if smContext.AccessType() == deregData.AccessType {
+				var problemDetails *models.ProblemDetails
+				problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
```

Comment:
`SendReleaseSmContextRequest()` needs to read the attribute of the `ue`, it will occur the concurrent read/write if another ngap/nas message for the same ue is sent to AMF.
Please help to check do we need a mutex lock here.

Reply:
Moreover, it seems like the behavior of AMF in this case is the same with `gmm_common.RemoveAmfUe(amfUe, false)`.
@iamelisahi @YouShengLiu

Reply:
> SendReleaseSmContextRequest() needs to read the attribute of the ue, it will occur the concurrent read/write if another ngap/nas message for the same ue is sent to AMF.
Please help to check do we need a mutex lock here.

`SendReleaseSmContextRequest()` will only use the ue.TimeZone and smContext.
Since the smContext is stored in ue.SmContextList and protected under the sync.Map, we can leave it alone.
However, the ue.TimeZone is not protected, it may need mutex lock to protect.

Reply:
> Moreover, it seems like the behavior of AMF in this case is the same with gmm_common.RemoveAmfUe(amfUe, false).

Yes, it is very similar but there are still some differences.

According to the TS 29.503 - Table 6.2.6.2.7-1: Definition of type Amf3GppAccessRegistrationModification - backupAmfInfo:
> This IE shall be included if the NF service consumer is an AMF and the AMF supports the AMF management without UDSF for the Modification of the BackupAmfInfo.
> The UDM uses this attribute to do an NRF query in order to invoke later services in a backup AMF, e.g. Namf_EventExposure

We think we should provide the BackupAmfInfo but the `UeCmDeregistration()` didn't.
Should we need to modify `UeCmDeregistration()`, e.g. add one more parameter to decide if it needs to put the BackupAmfInfo into Amf3GppAccessRegistrationModification?

Reply:
@YouShengLiu 

> Yes, it is very similar but there are still some differences.

Is it possible to reuse the duplicated part?

> We think we should provide the BackupAmfInfo but the UeCmDeregistration() didn't.
Should we need to modify UeCmDeregistration(), e.g. add one more parameter to decide if it needs to put the BackupAmfInfo into Amf3GppAccessRegistrationModification?

I think you are right, please help to fill the BackupAmfInfo.

Reply:
Does spec mention to use old amf as the backup amf? I think it's strange to notify UDM that the backup amf of e.g. amf1 is amf1 itself😂 Backup AMF may be a pre-config field to this AMF's config in my opnion.

Reply:
Spec didn't mention using old AMF as the backup AMF directly. I was thinking that the old AMF could be one of the backup AMFs when the new AMF is serving the UE.
I checked the data stored in the database, the serving AMF was the new AMF, and the old AMF became the backup AMF:
```json
{
  "_id": {
    "$oid": "650059c26b8b38f7f8ab93dc"
  },
  "ueId": "imsi-2089300007487",
  "amfInstanceId": "faf609ad-2cbd-4082-bf41-7a554e4aa626",
  "imsVoPs": "HOMOGENEOUS_NON_SUPPORT",
  "deregCallbackUri": "http://127.0.0.19:8000/namf-callback/v1/amf-implicit-deregistration/imsi-2089300007487",
  "initialRegistrationInd": true,
  "guami": {
    "amfId": "cafe01",
    "plmnId": {
      "mcc": "208",
      "mnc": "93"
    }
  },
  "ratType": "",
  "backupAmfInfo": [
    {
      "backupAmf": "AMF",
      "guamiList": [
        {
          "amfId": "cafe00",
          "plmnId": {
            "mcc": "208",
            "mnc": "93"
          }
        }
      ]
    }
  ],
  "purgeFlag": true
}
```
> I think it's strange to notify UDM that the backup amf of e.g. amf1 is amf1 itself😂 Backup AMF may be a pre-config field to this AMF's config in my opinion.
I agree that using pre-config is a better solution.

Reply:
Now there is no config for backup amf, so don't fill backup amf info to UDM for now.